File: uniswap_v3_math/full_math.fs
module uniswap_v3_math.full_math

open System.Numerics
open CustomInt256
open uniswap_v3_math.utils


let mul_div(a: UInt256, b: UInt256, denominator: UInt256): Result<UInt256, UniswapV3Error> =
    
    // 512-bit multiply [prod1 prod0] = a * b
    // Compute the product mod 2**256 and mod 2**256 - 1
    // then use the Chinese Remainder Theorem to reconstruct
    // the 512 bit result. The result is stored in two 256
    // variables such that product = prod * 2**256 + prod0
    let mutable outref = UInt256.MinValue
    let modV = UInt256.MaxValue
    let mm = a.MultiplyMod(&b, &modV, &outref)
    ()

File: UniswapV3/FullMath.cs
using System.Numerics;
using CustomInt256;

namespace UniswapV3;

public static class FullMath
{
    private static BigInteger? _UInt512Max {
        get;
        set;
    }

    private static BigInteger UInt512Max {
        get
        {
            if (_UInt512Max != null)
            {
                return _UInt512Max.Value;
            }

            UInt256.MaxValue.Convert(out var t);
            _UInt512Max = t * 2;
            return _UInt512Max.Value;
        }
    }
    
    public static UInt256 MulDiv(UInt256 a, UInt256 b, UInt256 denominator)
    {
        var a2 = (BigInteger)a;
        var b2 = (BigInteger)b;
        var denom2 = (BigInteger)denominator;

        var result = a2 * b2 / denom2;
        if (!(result <= (BigInteger)UInt256.MaxValue))
        {
            throw new OverflowException($"Result must be less than uint256 max {result.ToString()}");
        }

        return new UInt256(
            result.ToBytes32(true),
            true
        );
    }

    public static UInt256 MulDivRoundingUp(UInt256 a, UInt256 b, UInt256 denominator)
    {
        var result = MulDiv(a, b, denominator);
        a.MultiplyMod(b, denominator, out var mod);
        if (mod > 0)
        {
            if (!(result < UInt256.MaxValue))
            {
                throw new InvalidDataException("Foo");
            }
            result++;
        }

        return result;
    }
}